packages/react-reconciler/src/ReactFiberHooks.new.js

```diff
@@ -395,6 +395,22 @@ export function renderWithHooks<Props, SecondArg>(
       current !== null && current.type !== workInProgress.type;
   }
 
+  // Reset the memoCache index and create a backup copy in case of a setState during render
+  // or error, either of which can leave the cache in an inconsistent state
+  const memoCache = workInProgress.memoCache;
+  let previousMemoCache = null;
+  if (memoCache !== null) {
+    previousMemoCache = memoCache.data.map(array => array.slice());
+    if (workInProgress.alternate != null) {
+      // Backup to the alternate fiber in case the component throws
+      workInProgress.alternate.memoCache = {
+        data: previousMemoCache,
+        index: 0,
+      };
+    }
+    memoCache.index = 0;
+  }
+
```

Comment:
I think you can eliminate much of this code and similar code in other places by using the built-in hooks infra which can manage resets.

When first mounting you can create the memoCache and on each update copy it. If there is a renderPhaseUpdate the cache will get recreated. Additionally in dev mode with double-rendering of function components it will also get recreated (which is I think what you want in every case). It'd look something like this...

```
function mountMemoCache(size: number): Array<any> {
  const hook = mountWorkInProgressHook();
  hook.memoizedState = new Array(size);
  return hook.memoizedState;
}

function updateMemoCache(size: number): Array<any> {
  const hook = updateWorkInProgressHook();
  const current: Hook = (currentHook: any);
  hook.memoizedState = current.memoizedState.slice();
  return hook.memoizedState;
}
```

adding arg checks and warnings as appropriate.

This also has the nice benefit that it supports many memoCaches per Function component without any special index tracking

LMK if you want to talk through it. Also, maybe I don't fully understand the desired mount/update/re-render semantics so take my recc with a grain of salt :)